### PR TITLE
separate C/Fortran/Python packages

### LINKED ISSUE #15: fortran bindings for OSX?
I notice that the fortran bindings are currently explicitly disabled for OSX in
https://github.com/conda-forge/libxc-feedstock/blob/258cfdc26eb19b4605e92d03de340224823e95f8/recipe/build.sh#L3
This currently prevents me from creating a libxc-enabled cp2k conda-package for OSX.

I guess there is a reason for this - what happens to be the problem?

### PR DESCRIPTION
**Pull request by loriab**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] ~Reset the build number to `0` (if the version changed)~
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

closes #15

### CODE REVIEW

**Comment by loriab** [comment]
@ltalirz, @jan-janssen, @bjodah, here's another set of changes to consider (addresses #15)

- [x] separates the libxc packaging into 3 packages: basic C library, Fortran libraries (depends on C lib), and Python module (depends on C lib).
   - the main advantage here is that C consumers can avoid fortran library dependencies (libgfortran4 vs libgfortran5 has been known to cause trouble mixing channels). 
   - also, every package won't depend on python -- there should be one C, one F, and nPy Py packages per arch
- new capabilities:
   - [x] F for Mac 
   - [ ]  Py for Win (I don't think it was working before, but I think it's close to working now)
- big questions are:
   - [ ] do we want this split?
   - [ ] what should packages be named? at the moment, they're called `libxc-c`, `libxc-f`, & `pylibxc` to be very clear. I suggest they be called `libxc`, `libxc-f`, & `pylibxc`. however, there's also a case (for linux continuity) to leave the names as-is and add a metapackage `libxc` that pulls in all three.

### Pre-PR

package | libxc
---- | ----
Linux | C/F/Py
Mac | C/--/Py
Win | C/--/Py?

### post-PR

package | libxc-c | libxc-f | pylibxc
---- | ---- | ---- | ----
Linux | C | F | Py
Mac | C | F | Py
Win | C | -- | Py

TODO:
- [ ] turn other pythons and arch back on
- [ ] turn fctl deriv back up to 4e

**Comment by ltalirz** [comment]
Hi @loriab, thanks for asking!

> do we want this split?

I follow your argument behind splitting off the python interface into `pylibxc`. As for splitting the C and Fortran interface, I would personally be happy to keep both inside `libxc` as I guess it has no significant effect on package size / dependencies (or does it?).

**Comment by loriab** [comment]
Ok, new complication, conda-build gets confused over my intent if I rename the C library package to plain `libxc`. It complains about missing Fortran compiler, so my suspicion is that it's confusing the outer `requirements` provisioning section with the inner dependency-computing ones. Sometimes it's better to let the tools win, so my preferred names line in the table below, "recipe hates", is out.

I surveyed Debian and Fedora packaging combinations. They place C and Fortran libs together, as @ltalirz suggests. I'd like to have a way to request C alone so that libgfortran isn't pulled in, but I don't mind that way being called `libxc-c` rather than `libxc` itself. It remains to be tested whether the recipe will tolerate "recipe ok with?" line below. I don't want the plain libraries to have python depenence.


final names | ?-1 | ?-2 (dep on C pkg) | ?-3 (dep on C pkg)
---- | ---- | ---- | ----
Linux | C | F | Py
Mac | C | F | Py
Win | C | -- | Py
naming |
Debian | libxc | libxc | --
Fedora | libxc | libxc | python3-libxc
c-f pre-PR | libxc | libxc (mixed avail) | libxc (mixed avail)
recipe likes | libxc-c | libxc-f | pylibxc
recipe hates | libxc | libxc-f | pylibxc
@ltalirz sugg | libxc | libxc | pylibxc
recipe ok with? | libxc-c | libxc | pylibxc

**Comment by ltalirz** [comment]
No objections against option D from my side

**Comment by jan-janssen** [comment]
Maybe rather than `libxc-c` we could use something like `libxc-base` or `libxc-core` as that is also used by many other libraries.

**Comment by loriab** [comment]
fwiw, I think this is ready after the naming decision. (I'll need to rebuild with final names, high deriv, and correct build number. I propose waiting on #42 to see how the migrator bot deals with a split-output recipe.)

* the case for option D (libxc-c/libxc/pylibxc)
  * pro D: consistent w/ c-f Linux, Fedora, Debian
  * con E: inconsistent with above
  * on `libxc-c` vs `libxc-core`, the former seems more explicit.
* the case for option E (libxc/libxc-fortran/pylibxc)
  * pro E: C library is the heart of the project. consistent with c-f Mac & c-f Windows. 
  * con D: false advertising for c-f Windows that `libxc` always contains C & Fortran libs
  * the Fedora argument for putting C & Fortran libs together is that even consumers of the C lib are going to make use of the extra math and fortran dependencies of the Fortran lib as soon as the consumer pulls in LAPACK. With conda's multiple backends, that's not necessarily the case.

Anyways, I prefer E but am content with D. I don't want to prolong this thread, but I don't want to make unilateral decisions either. So perhaps folks can post/repost (1) whether you want a package split and (2) preference/tolerance for options D or E, then this can be finalized, discarded, or reworked at the weekend?

**Comment by ltalirz** [comment]
Cheers!

> (1) whether you want a package split and (2) preference/tolerance for options D or E

I'm sticking with my preference for option D. From a usability perspective, I think consistency with other distros is a significant plus, as is sparing users of the C/Fortran interface the need to think about which subpackage to use (they will try `libxc` and it will simply work).

I'm not familiar with the Fortran situation on Windows, but I guess if that ever becomes available, it can be added there as well? 

That said, I care mostly about having the Fortran interface available at all, and as long as things are clearly documented any of the options D-F would work for me.
Thanks for adding it @loriab !

**Comment by jan-janssen** [comment]
I am fine with option D as well.

**Comment by loriab** [comment]
Ok, option D it is; thanks for the feedback. I think this is ready for review and merge.

File: recipe/build.sh
Comment (ltalirz) [inline recipe/build.sh:3]: I guess this will simply be ignored if there is no fortran compiler around?

**Review by ltalirz — APPROVED** [review]
thanks @loriab!

File: recipe/build.sh
Comment (loriab) [reply recipe/build.sh:3]: It should error if there's no fortran compiler around for mac. compiler presence is assured by new lines 21 and 91. The reason it doesn't complain for Windows is the build/skip line for the fortran output and ENABLE_FORTRAN isn't turned on in bld.bat.

Comment (loriab) [reply recipe/build.sh:3]: In build.sh, the aarch and other  branches of the if/else could be merged, I think,  and just have a specialized value for derivative level

Comment (loriab) [reply recipe/build.sh:3]: Windows won't have a libxc package that gets updated (until we someday add a Windows fortran compiler build). But in practice, folks will continue to get the C-lib-only v6.1 package currently published when they request libxc win-64.

Comment (loriab) [reply recipe/build.sh:3]: Sorry for the barrage -- I keep thinking of additional aspects that might address your question. 

To my understanding, the outer level build/host requirements (before the outputs field) are what's present for the conda-build. So that has the merged requirements so that build.sh/bld.bat see C compilers, Fortran compilers, and Python, and the line you highlighted for Mac can be fulfilled. The package builds once with everything (C/F/Py) enabled. Then the files field for each output name divides the spoils into packages. Then the output name build/host/run requirements are processed into the dependencies to be assigned to each output; so libxc-c depends on libgcc, libxc depends on libgcc and libgfortran, and pylibxc depends on libgcc and the build-time python.

Comment (ltalirz) [reply recipe/build.sh:3]: Thanks @loriab for this clear explanation, indeed I have/had not yet quite grasped how multiple outputs work under the hood.

Just one follow-up question: if it's the outer level build requirements that define the environment at build time, do we still need `build` requirements in the output packages?
I get that you still want `host`/`run` requirements, but I guess the `build` requirements then don't do anything?

Comment (ltalirz) [reply recipe/build.sh:3]: I read [here](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#subpackage-requirements) that the `build` requirements of the subpackage are available during the "packaging stage" of that subpackage (perhaps some scripts you want to run to postprocess the `files`).

As far as I understand the docs, it seems to me that in our case both the `build` and `host` requirements of the outputs can be dropped

Comment (ltalirz) [reply recipe/build.sh:3]: P.S. If you need some dependencies for the `test` stage of the outputs, there is also
```
test:
  requires:
    - package
```

Comment (loriab) [reply recipe/build.sh:3]: > I get that you still want host/run requirements on the output packages, but I guess the build requirements then don't do anything (and might, over time, become outdated vs the top-level build requirements)?

It does reason that build/host are unneeded, but this is where `run_exports` is doing its work. If libxc had needed MKL to compile, we'd have put `mkl-devel` in the host. Because that mkl recipe has a run_exports, placing `mkl-devel` in the libxc recipe's host section would have effectively imposed a properly pinned runtime mkl dependence upon libxc. That's the way most run_exports (incl. the build/run_exports ones in this recipe) work. There's another type: "strong" run_exports that jump two sections from build to run. They're responsible for compiler('c') imposing libgcc runtime dependence. 

So as far as I know, of outer build/host/run and inner build/host/run, outer run are the only section that can be skipped.

**Comment by loriab** [comment]
@conda-forge-admin, please rerender

**Comment by awvwgk** [comment]
Would it make sense to keep a `noarch: generic` libxc package around which provides the same functionality as previous builds to avoid packages currently depending libxc breaking after this build is pushed (possibly packages depending on the Python bindings)?

**Comment by loriab** [comment]
> Would it make sense to keep a noarch: generic libxc output around which provides the same functionality as previous builds to avoid packages currently depending libxc breaking after this build is pushed (possibly packages depending on the Python bindings)?

Hmm, thanks @awvwgk, I hadn't considered that, as I didn't think the py bindings were much used. Here's a search of libxc on c-f: https://github.com/search?q=org%3Aconda-forge+libxc&type=code . Below is a quick survey, and it doesn't look like pylibxc is being used. Does anyone know of a package that would be inconvenienced by the missing Py module?

* pyscf is pinning to a libxc v5.1 (https://github.com/conda-forge/pyscf-feedstock/blob/main/recipe/meta.yaml#L38) so they could we warned to update the recipe when they move to v6
* bigdft-suite is pinning to <v5 (https://github.com/conda-forge/bigdft-suite-feedstock/blob/main/recipe/meta.yaml#L44)
* skprogs isn't depending on python, so I don't think they're affected (https://github.com/conda-forge/skprogs-feedstock/blob/main/recipe/meta.yaml#L26)
* gpaw I'm not sure, but it looks like libxc is going into a setup.py extension module, so it's using the library, not the pymod (https://github.com/conda-forge/gpaw-feedstock/blob/main/recipe/build.sh#L39)
* sirius is using unversioned libxc (https://github.com/conda-forge/sirius-feedstock/blob/main/recipe/meta.yaml#L33), but it's sending it to cmake, so probably using the library
* skprogs is also using unversioned libxc (https://github.com/conda-forge/skprogs-feedstock/blob/main/recipe/meta.yaml#L26) but there's a patch to remove Python, and cmake is invoked, so probably using the library
* abinit is pinning to <v5 (https://github.com/conda-forge/abinit-feedstock/blob/main/recipe/meta.yaml#L39)

**Comment by awvwgk** [comment]
The Python package seems to be almost pure Python, is there a possibility to generate it as `noarch: python` or if not have the Python matrix appear only in the *pylibxc* output? Currently the matrix seems to build the C and Fortran part for each Python version which is a bit wasteful in terms of CI.

**Comment by loriab** [comment]
> The Python package seems to be almost pure Python, is there a possibility to generate it as noarch: python or if not have the Python matrix appear only in the pylibxc output? Currently the matrix seems to build the C and Fortran part for each Python version which is a bit wasteful in terms of CI.

Definitely it's wasteful of CI, but I've sometimes found it futile to cut down on builds in conda recipes and get the hashes and deps to compute correctly. Nevertheless, I found tomopy as a model, and there are hopeful signs in my test builds.

DO NOT MERGE at present, please.

**Comment by loriab** [comment]
@conda-forge-admin, please rerender

**Comment by loriab** [comment]
Thanks, @awvwgk, that worked and reduced the number of library builds from 24 to 4! `noarch` might be possible with a dummy file and special handling of Windows. But this at least cuts out the duplicate builds, and I think Susi has plans to redo pylibxc later anyways.

CI is re-running now with the high derivatives. After that, it should be good to merge again.

File: recipe/build.sh
Comment (loriab) [reply recipe/build.sh:3]: @ltalirz, revised understanding of outer/inner requirements based on the build reduction work below: split output packages can each have their own subpackage build script, or a single build script for all (situation above), or a mix (situation below). If each subpackage has its own build script, no outer requirements block is necessary. In the mixed case, the output requirements covers the subpackages built with the common build script, then the "files" field partitions them. So for the below, the outer requirements/build has c & fortran compilers so that packages libxc-c and libxc (Fortran) can be build in one pass per arch with the common build.sh/bld.bat. The python dep has been removed from the outer and confined to the inner pylibxc requirements/host so that each arch iterates over pythons and runs the trivial build_py.sh/bld_py.sh.

**Comment by loriab** [comment]
@jan-janssen @bjodah, I think this is ready to go when you're content with it. Please let me know any questions or concerns or if you'd simply prefer to put it on hold for a while. PRs #42 and #44 can plan their build numbers accordingly.

**Pull request merged**
